### PR TITLE
Cancellation token support

### PR DESCRIPTION
**Pull request by dVakulen**
#1516; 
Made through inspecting grain method arguments, wrapping found cancellation tokens and subscribing cancellation of tokens on crossing domain bounds. 
Added costs for general case - two grain method arguments array iterations with type checking. 
For consumer in case of in-silo token passing there's cost of 2 wrappers allocations, wich is neglectable. For remote case - mostly it will be the same as if user cancelled remote `TaskCompetitionSource` by himself: 1 grain call, but when token gets cancelled after grain call message serialization and before its deserialization there will be 2 grain calls (token holder grain will be allocated on different silo).

### CODE REVIEW

**Comment by gabikliot** [comment]
You would probably want to write an explanation of the code: how it works, what is `CancellationTokenHolderGrain `, what is GCObserver, why do we need the wrappers...

I don't know if/how this is related, but I also found it very hard to follow: https://github.com/dotnet/orleans/issues/1516#issuecomment-193493372.

**Comment by dVakulen** [comment]
Tests fails doesn't seems to be related to changes in this PR.

**Comment by dVakulen** [comment]
@gabikliot Indeed, I intended to write explanations as comments in PR, and in case of approve - commit them.

File: src/Orleans/Threading/CancellationTokenManager.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+        private static void RegisterTokenCallbacks(CancellationToken ct, Guid tokenId)
+        {
+            if (!ct.CanBeCanceled) return;
+            ct.Register(
+               async (f) =>
+               {
+                   await Cancel(tokenId);
+               },
+                new GCObserver(
```
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenManager.cs]: GCObserver is essentially an instance that will be collected with `TaskCompletionSource` that issued the `CancellationToken`, so that linked `TaskCompletionSource` on different silo can be freed aswell

**Comment by gabikliot** [comment]
I think a design in one place would be easier to follow, read, discuss and preserve in the issue.

File: src/Orleans/Threading/CancellationTokenWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.Threading;
+
+namespace Orleans.Threading
+{
+    /// <summary>
+    /// Used as replacement of CancellationToken during network roundtrips
+    /// </summary>
+    [Serializable]
+    internal class CancellationTokenWrapper
```
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenWrapper.cs]: Wraps the cancellation token, so that on passing though serialization process linked `TaskCompletionSource` can be created. An alternative was to handle serialization of  `CancellationToken` directly in `BuiltInTypes`, but the thread that performs message handling and serialization have `RuntimeContext.Current`, which is essential for making grain calls, set to null.

**Comment by dVakulen** [comment]
Lifetime of the `CancellationToken` passed to the grain method: on sening the request it's being wrapped in `CancellationTokenWrapper`. If request is local - before invoking of actual grain method it's just being unwrapped. For the remote case: on message serialization creates subscription on token cancel event,  which will cancel linked `TaskCompletionSource` located in the `CancellationTokenHolderGrain`; after deserialization of the wrapper the abovementioned grain will be allocated on same silo with called grain, and wrapper will be replaced with actual token from holder grain. If `CancellationTokenHolderGrain` cancellation method called before grain's allocation there is a possibility that it will be allocated on another silo than target, which will lead to another token's roundtrip.

**Comment by dVakulen** [comment]
Synced in order to re-run the tests.

File: src/Orleans/Threading/CancellationTokenManager.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        private static async Task Cancel(Guid tokenID)
+        {
+            await RuntimeClient.Current.InternalGrainFactory.GetGrain<ICancellationTokenHolderGrain>(tokenID).Cancel();
+        }
+
+        private static async Task Dispose(Guid tokenID)
+        {
+            await RuntimeClient.Current.InternalGrainFactory.GetGrain<ICancellationTokenHolderGrain>(tokenID).Dispose(); ;
```
Comment (ReubenBond) [inline src/Orleans/Threading/CancellationTokenManager.cs]: nit: superfluous `;`

File: src/Orleans/Threading/CancellationTokenManager.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        public static Task<CancellationToken> GetCancellationToken(Guid tokenId)
+        {
+            return RuntimeClient.Current.InternalGrainFactory.GetGrain<ICancellationTokenHolderGrain>(tokenId).GetCancellationToken();
+        }
+
+        private static void RegisterTokenCallbacks(CancellationToken ct, Guid tokenId)
+        {
+            if (!ct.CanBeCanceled) return;
+            ct.Register(
+               async (f) =>
```
Comment (ReubenBond) [inline src/Orleans/Threading/CancellationTokenManager.cs]: Just to make sure it's absolutely clear: this `async` func here is unobserved, unawaited, and is being seen by the `CancellationToken.Register` method as an `Action<object>`. If that's the intention and the implications have been carefully considered, then could we mark this with a comment discussing that? It just makes me nervous.

Nit: typically single parameter lambdas omit the superfluous parenthesis.

File: src/Orleans/Threading/CancellationTokenManager.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        public static CancellationTokenWrapper WrapCancellationToken(CancellationToken ct)
+        {
+            var tokenId = Guid.NewGuid();
+            return new CancellationTokenWrapper(tokenId, ct)
+            {
+                OnSerialization = () =>
+                {
+                    RegisterTokenCallbacks(ct, tokenId);
```
Comment (ReubenBond) [inline src/Orleans/Threading/CancellationTokenManager.cs]: It's often difficult to judge where allocations will occur. I believe that you can avoid allocating a closure here by using `this.CancellationToken` & `this.Id`. You could alternatively accept an `Action<CancellationToken, Guid>` parameter in the constructor.

Someone please correct me if I'm wrong.

File: src/Orleans/Threading/CancellationTokenManager.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                   await Cancel(tokenId);
+               },
+                new GCObserver(
+                    () =>
+                    {
+                        Dispose(tokenId).Ignore();
+                    }));
+        }
+
+        private static async Task Cancel(Guid tokenID)
```
Comment (ReubenBond) [inline src/Orleans/Threading/CancellationTokenManager.cs]: nit: inconsistent naming `tokenId` vs `tokenID`. Also below.

File: src/Orleans/Threading/CancellationTokenHolderGrain.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+        public Task<CancellationToken> GetCancellationToken()
+        {
+            return Task.FromResult(cancellationTokenSource.Token);
+        }
+
+        public Task Cancel()
+        {
+            cancellationTokenSource.Cancel();
+            DelayDeactivation(TimeSpan.FromSeconds(30));
```
Comment (ReubenBond) [inline src/Orleans/Threading/CancellationTokenHolderGrain.cs]: If this number was something tiny like 0.1ms, would this be unreliable?

**Comment by ReubenBond** [comment]
LGTM once the comments are addressed. I'm also a little uncertain about the use of a grain to handle cancellation - possibly a `ConcurrentDictionary<Guid, WeakReference<CancellationTokenWrapper>>` would be better. I haven't explored that option, but perhaps you have?

File: src/Orleans/Threading/CancellationTokenManager.cs (hunk 4 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenManager.cs]: It's unawaited on purpose, but good point on it being unobserved, will add continuation with exception logging.

File: src/Orleans/Threading/CancellationTokenHolderGrain.cs (hunk 7 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenHolderGrain.cs]: Will add additional number to the https://github.com/dotnet/orleans/pull/1569/files#diff-7ddbc03b5e6c00821897b355928dda00R14

**Comment by dVakulen** [comment]
@ReubenBond The issues with `ConcurrentDictionary<Guid, WeakReference<CancellationTokenWrapper>>`:
-  weak reference means that token wrapper can be collected before cancellation request arrives; 
- the main problem is that finding out on which silo target grain is located would require putting call to the [address message](https://github.com/dotnet/orleans/blob/7113903a3f5b41ba722ef18148f3c38ef7a37f42/src/OrleansRuntime/Core/Dispatcher.cs#L531) method from `Dispatcher` into `InsideGrainClient`, forwarding cancellation request in case of original message being forwarded, and bringing the management of lifecycle of the `CancellationTokenWrapper` on our shoulders. Are you having suspicion that solution with hoder grain will be inefficient compared to the `dictionary` one? 

About better solution - I've tried to handle serialization of `CancellationToken` directly in `BuiltInTypes`, which would've allow handling tokens not only in signatures of grain methods, but also as properties of custom objects, but faced issues described in https://github.com/dotnet/orleans/pull/1569#discussion_r56281832.

File: src/Orleans/Threading/CancellationTokenHolderGrain.cs (hunk 7 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenHolderGrain.cs]: And no, it would not. The cancellation wil either happen before the token has passed throuht the serialization, and already cancelled one will be deserialized, or slightly after, in which case the cancel request will be issued before token's deserialization, the holder grain will be allocated, and after deserialization the token will be taken from that grain.

File: src/Orleans/Threading/CancellationTokenManager.cs (hunk 6 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenManager.cs]: Fixed

File: src/Orleans/Threading/CancellationTokenManager.cs (hunk 5 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenManager.cs]: Fixed

File: src/Orleans/Threading/CancellationTokenManager.cs (hunk 3 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenManager.cs]: Fixed

File: src/Orleans/Threading/CancellationTokenHolderGrain.cs (hunk 7 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenHolderGrain.cs]: Sorry, thought about completly different number. About this one: in case of cansellation happened before token requesting this will be unreliable with 0.1 ms. 30 seconds should be sufficient for almost all cases, as default request timeout is 30 seconds too.

**Comment by veikkoeeva** [comment]
A few thoughts:
* Should we open a new issue about providing documentation for this feature? I could imagine it would have an example, rational and guarantees (or what can't be guaranteed) and resource implications if any (CPU, memory, grain call limits if any).
* Some of the rationale comments in this issue would make sense in the code too. Most people will read code afterwards. :)
* Considering future, this particular scenario might lend well into a local integration test wherein a self-osted Owin HTTP server is started on a random port on ephemeral range, which is then called with `HttpClient` to issue a long-standing query, which will then be cancelled. Both to provide an example of one intended use-case scenario and also keep the implementation bits honest in that we'd detect if any/some scenarios would become broken for a reason or another.

File: src/Orleans/Threading/CancellationTokenManager.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+                {
+                    if (task.IsFaulted)
+                    {
+                        var logger = TraceLogger.GetLogger("CancellationTokenManager", TraceLogger.LoggerType.Runtime);
+                        if (logger.IsWarning)
+                        {
+                            logger.Warn(ErrorCode.Runtime_Error_100332, "Remote token cancellation failed", task.Exception);
+                        }
+                    }
+                });
```
Comment (veikkoeeva) [inline src/Orleans/Threading/CancellationTokenManager.cs]: Would it make sense to use `TaskContinuationOptions.OnlyOnFaulted` here instead of explicit `task.Faulted`? I'm not entirely sure if it would improve inlining etc.

File: test/Tester/CancellationTests/CancellationTokenPassingTests.cs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+using System.Threading.Tasks;
+using Orleans;
+using Orleans.TestingHost;
+using UnitTests.GrainInterfaces;
+using Xunit;
+using UnitTests.Tester;
+
+namespace UnitTests.MembershipTests
+{
+    public class CancellationTokenPassingTests : HostedTestClusterPerTest
```
Comment (veikkoeeva) [inline test/Tester/CancellationTests/CancellationTokenPassingTests.cs]: For another non-native speaker this suggests the `CancellationToken` tests here should pass. Just a note. :)

File: test/Tester/CancellationTests/CancellationTokenPassingTests.cs [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+                StartFreshOrleans = true,
+                StartPrimary = true,
+                StartSecondary = true,
+                AdjustConfig = config =>
+                {
+                    config.Globals.DefaultPlacementStrategy = "ActivationCountBasedPlacement";
+                }
+            });
+        }
+
```
Comment (veikkoeeva) [inline test/Tester/CancellationTests/CancellationTokenPassingTests.cs]: What would happen if `CancellationToken`s were linked together with [CreateLinkedTokenSource ](https://msdn.microsoft.com/en-us/library/ee191553(v=vs.110).aspx), say, in a fan-out scenario? (cf. http://blogs.msdn.com/b/pfxteam/archive/2009/05/22/9635790.aspx).

File: src/Orleans/Threading/CancellationTokenHolderGrain.cs (hunk 7 above)
Comment (veikkoeeva) [inline src/Orleans/Threading/CancellationTokenHolderGrain.cs]: What would happen in the cases not covered in "almosta all"? What consequences would it have and should they bo documented?

Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenHolderGrain.cs]: I can imagine some edge case with extremely bad network which would requere multiple retries of message sending, that could span over 30 seconds,  with grain call timeout adjusted to more than 30 seconds. I could use the grain call timeout from the config as delay, but thought that probability of abovementioned case vanishingly low.

File: test/Tester/CancellationTests/CancellationTokenPassingTests.cs (hunk 9 above)
Comment (dVakulen) [inline test/Tester/CancellationTests/CancellationTokenPassingTests.cs]: Yes, it does for me too, but couldn't come with better name, that wouldn't be too long.

Comment (veikkoeeva) [inline test/Tester/CancellationTests/CancellationTokenPassingTests.cs]: As a corollary it feels also a tautology. :) OK, I lack ideas too, but so as not become complete empty, maybe just "CancellationTokenTests"?

File: src/Orleans/Threading/CancellationTokenHolderGrain.cs (hunk 7 above)
Comment (veikkoeeva) [inline src/Orleans/Threading/CancellationTokenHolderGrain.cs]: Would using the configured timeout be a lot of work? Say something like
```
var safetyCautionLimit = TimeSpan.FromSeconds(1);
DelayDeactivation(configuredDelay.Add(safetyCautionLimit));
```
I don't know what would the consequence here could be if that hardcoded 30 seconds would be off, so I my gut reaction might be just that.

File: test/Tester/CancellationTests/CancellationTokenPassingTests.cs (hunk 9 above)
Comment (dVakulen) [inline test/Tester/CancellationTests/CancellationTokenPassingTests.cs]: "CancellationTokenTests" seems like testing the cancellation token itself. But if it's just me - I can change to this name.

File: test/Tester/CancellationTests/CancellationTokenPassingTests.cs (hunk 10 above)
Comment (dVakulen) [inline test/Tester/CancellationTests/CancellationTokenPassingTests.cs]: That woun't change anything from this implementation perspective.

File: src/Orleans/Threading/CancellationTokenManager.cs (hunk 8 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenManager.cs]: Fixed

**Comment by dVakulen** [comment]
Added ability to configure delay before deactivation of the holder grain and merged.

File: src/Orleans/Threading/CancellationTokenHolderGrain.cs (hunk 7 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenHolderGrain.cs]: Hard reseted the addition of configurable delay, will add as separate PR later.

**Comment by veikkoeeva** [comment]
I have exhausted myself here, LGTM. :+1:

**Comment by dVakulen** [comment]
@dotnet-bot test this please

File: src/Orleans/Threading/CancellationTokenManager.cs [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            ct.Register((f) =>
+            {
+                Cancel(tokenId).ContinueWith(task =>
+                {
+                    var logger = TraceLogger.GetLogger("CancellationTokenManager", TraceLogger.LoggerType.Runtime);
+                    if (logger.IsWarning)
+                    {
+                        logger.Warn(ErrorCode.Runtime_Error_100332, "Remote token cancellation failed", task.Exception);
+                    }
+                }, TaskContinuationOptions.OnlyOnFaulted);
```
Comment (ReubenBond) [inline src/Orleans/Threading/CancellationTokenManager.cs]: Perhaps this exception logging code should sit inside of `Cancel` using a try/catch

**Comment by ReubenBond** [comment]
Great work! LGTM after 1 comment: https://github.com/dotnet/orleans/pull/1569/files#r56586711

File: src/Orleans/Threading/CancellationTokenManager.cs (hunk 11 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenManager.cs]: If it will be moved to the Cancel method the need to add async & await keywords to this lambda will arise. If its ok - I'll do it.

Comment (ReubenBond) [inline src/Orleans/Threading/CancellationTokenManager.cs]: The the continuation is not being observed either way. No need to actually add async/await, it wont affect anything. You may as well move the catch-all to `Cancel` and `.Ignore()` the call to `Cancel` here.

Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenManager.cs]: Done.

**Comment by dVakulen** [comment]
Rebased.

**Comment by dVakulen** [comment]
@ReubenBond, @veikkoeeva thank you for the review.

**Comment by ReubenBond** [comment]
Thank you, @dVakulen!

**Comment by gabikliot** [comment]
@ReubenBond , do you understand how it works? Can you please explain it to me? The whole design.

**Comment by ReubenBond** [comment]
Conceptually: intercept CancellationTokens as they cross silo boundaries (not if they are a part of a complex object, only if they are passed directly), replace them with a surrogate, and register a handler with the token which will fire a call to the remote silo upon cancellation.

Each token is associated with a unique id when its surrogate is created. When a surrogate is received, a `[PreferLocalPlacement]` grain is called to get the corresponding `CancellationToken` for that id, which is then swapped in for the argument/result. When cancellation is requested, a `Cancel()` is sent to that grain, which then cancels it's token & completes the process.

I didn't register that you weren't yet satisfied with the PR, sorry for that. I see now that there is a race condition or grain directory consistency issue. What's the best course of action, should I revert the commit?

**Comment by jdom** [comment]
Were perf tests ran before merging this? I'm just asking because I see that it touches on some hot/warm paths. I honestly didn't review this thoroughly yet, so maybe it's harmless, but I would be a lot more comfortable if we know that this went through a little bit of scale testing... we need to know that this doesn't affect silos not using this feature, and also if you use this feature that it doesn't fall down once you go beyond a simple demo without load.

**Comment by dVakulen** [comment]
@jdom About perf implications - type checking even for millions of iterations taking only tens to hundreds nanoseconds. So it definitely couldn't be seen at perf tests. As for possibility of falling down on load when using this feature - I've ran local test, in which the  [tests](https://github.com/dotnet/orleans/pull/1569/files#diff-7ddbc03b5e6c00821897b355928dda00R32) were run in parralel for 3 million iterations, silos were located on 2 machines.

**Comment by sergeybykov** [comment]
I have to admit I also was behind on this PR and planned to catch up this weekend. I just started perf tests to see if there's any regression there.

My immediate design question before I fully understand the code is about `CancellationTokenHolderGrain`. Wouldn't a system target be a better option instead?

**Comment by gabikliot** [comment]
I don't yet even understand the goal of this grain. What is it doing? What is its ID? Is it a one global grain?

**Comment by dVakulen** [comment]
The `CancellationTokenHolderGrain` contains `CancellationTokenSource` so that it can be cancelled remotely. 
I'n not that familiar with system target's characteristics, so I may be wrong, but I dont think that it would be a better option.

**Comment by sergeybykov** [comment]
As I understood, `CancellationTokenHolderGrain` is created for each cancellation token as a remotely callable endpoint. This brings quite a bit of overhead of the grains that get activated and linger around for a while plus additional pressure on the grain directory. A system target per silo as an endpoint for all tokens in the silo should be more efficient I think. If throughput of a single system target isn't enough (I doubt that), then we can easily create a pool of them like we do for pulling agents.

Did I misunderstand this aspect of the design?

**Comment by dVakulen** [comment]
@sergeybykov The understanding is correct. But how the cancel requests will be routed to the target silo in case of system target usage?

**Comment by sergeybykov** [comment]
The target silo address would have to be included in the token wrapper, I suppose.

Though I still haven't grasped the code. It looks like `InsideRuntimeClient.Invoke()` registers the cancellation endpoint by activating a `CancellationTokenHolderGrain` which is expected to be locally placed. There is obviously a potential race here, as the caller could have already called `Cancel()`, and thus activated the grain in its silo.

The system target approach that I thought about won't work this way 1:1 because the caller would need to know the silo where the call is being executed. Although considering the race in the PR, the complexity may end up being the same.

One option is for `InsideRuntimeClient.Invoke()` to make the local `CancellationTokenManager` make a call to the system target of the caller to register itself as the executing silo for the call. There would be a time gap during which there will be nowhere to send a cancellation message, but no race.

Another option I can think of is not to register executing silo at all, and send cancellations as a broadcast to all silos. Assuming that cancellations are rare, there will be no extra messages until a cancellation is requested but N messages for every cancellation.

**Comment by dVakulen** [comment]
The race condition is known and resulting in one more token holder grain placement.

**Comment by sergeybykov** [comment]
I am more concerned about the grain activation/registration cost that would be paid even in case of no cancellations. I think we better have zero cost unless cancellations are sent, even if the cost of a cancellation is higher.

**Comment by dVakulen** [comment]
Your first suggested option means non-zero cost of each cancellation sent, and the second probably  woun't scale at big deployments. The only way of implementing this with as low cost as possible is sending cancellation message which will reach target grain silo using already existing methods such as `AddressMessage` and message forwarding. Such implementation will require additional changes in core Orleans, but maybe can be useful in some another cases.

**Comment by sergeybykov** [comment]
We can easily make the broadcast option scale by batching cancellation calls, e.g. send not more that 10 batches a second. However, an app design that generates a significant rate of cancellations smells wrong to me. If cancellations are infrequent, then scalability of broadcasts isn't a concern I think.

**Comment by dVakulen** [comment]
Adn binding target cancellation token source to the target grain deactivation will remove the need of making the dispose call. Makes sense, will try to implement this approach with batching.

**Comment by dVakulen** [comment]
Meanwhile if the perf test woun't show any perf degradation I think there's no need to revert current implementation, as 70% of it can be reused.

**Comment by ReubenBond** [comment]
@dVakulen I think revert is the right way - I should not have merged it yet. Reverting it does not mean that we lose any of it, just that it can be iterated before the code is included in master. It still has the `work-in-progress` tag and we should document & discuss the design before it goes in.

File: src/Orleans/Threading/CancellationTokenHolderGrain.cs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    [PreferLocalPlacement]
+    internal class CancellationTokenHolderGrain : Grain, ICancellationTokenHolderGrain
+    {
+        private CancellationTokenSource _cancellationTokenSource;
+        private bool _collectionScheduled;
+        private TimeSpan _safetyCautionLimit = TimeSpan.FromSeconds(1);
+
+        public override Task OnActivateAsync()
+        {
+            DelayDeactivation(TimeSpan.FromDays(5));
```
Comment (gabikliot) [inline src/Orleans/Threading/CancellationTokenHolderGrain.cs]: 5 days? why? you expect the call to be alive and cancelable for 5 days?

**Comment by gabikliot** [comment]
OK, maybe I understand the source of the confusion now. @dVakulen tried to implement it on the "application layer" while minimizing runtime changes and this is why he created a grain to basically map the token to a grain to which this token was sent too. Makes sense now. But even if we went with this approach, we better create "token mapping service" as a small collection of grains. Having one grain per one token (meaning, one extra grain for EVERY call that passes cancellation token - grain per call) is very inefficient.
But you anyway had to a lot of runtime changes (inside grain client change is the major one), so I don't think its a goal anymore.
How about an alternative design:

Lets simplify first and look on one hop calls: client makes a grain call with token and then tries to cancel. All we need to do is to remember, on the client, where the token was sent to and send a msg there, to this grain, intercept it at the runtime (just around where grain call is made) and cancel the token. Basically, you need to remember a call context. We of course have it already - CallbackData. It holds the context of the grain call, so when reply arrives we can find the Task to resolve. So you just store the token there and upon cancel call need to be able to find that CallbackData locally. 
Then you make a call to the grain. Now you have 2 options: either use grain extensions: activation installs an extension when it gets a token 1st time, or we can just change grain base class to always implement internal interface `Cancel`.
So basically, in this solution, there is zero extra cost, zero extra state (just need to link from CallbackData  to a token and back).

Now lets look at multi hop - client calls into grain A with a token which passes this same token to grain B and now you want to cancel. My approach will work automatically, but the cancellation call will have to go via the silo that hosts A. If this silo crashed, we wont be able to send the token to B. First question here is do we even want to guarantee that even in case of failures all calls are cancelable? maybe not? If yes, then indeed we need to think in terms of more complicated solution of "token mapping service" and reason about failures and all this stuff. Your initial solution does not deal with failures either I think (your holder token grain silo can fail too). Sergey's proposal with broadcast does work for this case.

Another question which we need to start with is prog. model. Do we want to support standard cancellation token as is? The problem is what if the grain defined the callback on the token, OnCancelled? You need to make sure to invoke it, but not violate singe threaded execution and also deal with non-re-entrant grains. Adds much complexity. An alternative is to introduce OrleansCancellationToken, which mimics the standard one, but does not have OnCancelled. Ugly to define our won, but simplifies a lot.

Finally, to close, @dVakulen , this is what I meant when I asked for the design. **When we implement new feature, everyone will have their one opinion**. Just be prepared for that. We have to look deep in different options and discuss them. And better not even in GH on the issue, as it is really not a good forum for design. A better way is to write a word/one note doc (office online is free). Sergey can probably create a folder for Orleans designs. Or I can create a Google doc for you

**Comment by sergeybykov** [comment]
I agree we better revert this one. Useful code won't disappear anyway. It was a nice attempt, but let's step back and design a more scalable and performant solution.

@dVakulen Thank you for taking the lead on this!

**Comment by sergeybykov** [comment]
I agree we better revert this one. Useful code won't disappear anyway. It was a nice attempt, but let's step back and design a more scalable and performant solution.

@dVakulen Thank you for taking the lead on this!

**Comment by dVakulen** [comment]
Thanks for the offer @ReubenBond, but I'll try to handle this one by myself. I think we can implement together something bigger than this.

**Comment by dVakulen** [comment]
@gabikliot  > . A better way is to write a word/one note doc (office online is free). Sergey can probably create a folder for Orleans designs 
Yep, this is better idea than marking issues with some sort of `design` label, but in issues it's easier to discuss the design, so I suggest to move to the online docs only final designs.

**Comment by jdom** [comment]
I successfully used Word online and it's commenting features for online collaboration. It's easy to have focused discussions there,  as it's more in line than an issue thread.

**Comment by dVakulen** [comment]
Then the folder in Word online for Orleans designs is definitely needed. cc @sergeybykov.

File: src/Orleans/Threading/CancellationTokenHolderGrain.cs (hunk 12 above)
Comment (dVakulen) [inline src/Orleans/Threading/CancellationTokenHolderGrain.cs]: I've seen tasks that were running for more than 20 hours. 5 days is just to be on the safe side..
